src/e2e_tests.rs

```diff
@@ -0,0 +1,649 @@
+    let processor_handler = tokio::spawn(processor(id, storage.clone()));
+
+    let api_handler =
+        tokio::spawn(HttpApi::new(id, address, storage.clone(), ident.clone(), db.clone())?.run());
+
+    let proxy_handler = tokio::spawn(ProxyWorker::new(id, 10, storage.clone(), db.clone()).run());
+
+    let workers_handler = tokio::task::spawn(HttpWorker::new(1000, storage, db, ident).run());
+
+    tokio::select! {
+        result = processor_handler => {
+            if let Err(err) = result {
+                bail!("message processor panicked unexpectedly: {}", err);
+            }
+        },
+        result = api_handler => {
+            match result {
+                Err(err) => bail!("http server panicked unexpectedly: {}", err),
+                Ok(Ok(_)) => bail!("http server exited unexpectedly"),
+                Ok(Err(err)) => bail!("http server exited with error: {}", err),
+            }
+        },
+        result = workers_handler => {
+            if let Err(err) = result {
+                bail!("http workers panicked unexpectedly: {}", err);
+            }
+        },
+        result = proxy_handler => {
+            if let Err(err) = result {
+                bail!("proxy workers panicked unexpectedly: {}", err);
+            }
+        },
+    };
+    unreachable!();
+}
+enum KPair {
+    EdPair,
+    SrPair,
+}
+
+fn create_test_signer(p: KPair) -> Result<impl Signer> {
+    match p {
+        KPair::EdPair => {
+            let (_pair, string, _seed) = EdPair::generate_with_phrase(None);
+            let ed_signer = identity::Ed25519Signer::try_from(string.as_ref()).unwrap();
+            Ok(ed_signer)
+        }
+        KPair::SrPair => {
+            let (_pair, string, _seed) = SrPair::generate_with_phrase(None);
+            let sr_signer = identity::Ed25519Signer::try_from(string.as_ref()).unwrap();
+            Ok(sr_signer)
+        }
+    }
+}
+
+async fn create_local_redis_storage(port: usize) -> Result<impl Storage + ProxyStorage> {
+    let redis_url = format!("redis://localhost:{}", port);
+    let pool = redis::pool(redis_url)
+        .await
+        .context("unable to create redis connection")?;
+    let storage = RedisStorage::builder(pool).build();
+    Ok(storage)
+}
+
+fn create_mock_db(twins: Vec<&Twin>) -> Result<impl TwinDB + Clone> {
+    let mut db = InMemoryDB::default();
+
+    for twin in twins {
+        db.add(twin.clone());
+    }
+    Ok(db)
+}
+
+fn start_redis_server(port: usize) -> Result<()> {
+    Command::new("sh")
+        .arg("-c")
+        .arg(format!(
+            "docker run --name redis-test-{} --rm -d -p {}:6379 redis",
+            port, port
+        ))
+        .output()
+        .expect("failed to execute process");
+    Ok(())
+}
+
+fn stop_redis_server(port: usize) -> Result<()> {
+    Command::new("sh")
+        .arg("-c")
+        .arg(format!("docker stop redis-test-{}", port))
+        .output()
+        .expect("failed to execute process");
```

Comment:
for this one, indeed Result is useless. I would/will instead change the return type and make it a void function because there is no need (here) to propagate the error, I want it to panic so the test fails and I have a verbose output where and why it panicked exactly without following the code to see from where the error was propagated. please, resolve this if you agree.